src/HW_models/nrfx_bsim_redef.h

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2018 Oticon A/S
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file redefines macros from nrfx that are not to be used
+ * with simulatoed peripherals.
+ */
+
+#ifndef NRF_SOC_IF_H
+#define NRF_SOC_IF_H
+
+#include <stddef.h>
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ * Redefine the base addresses.
+ */
+extern NRF_AAR_Type NRF_AAR_regs;
+#undef NRF_AAR_BASE
+#define NRF_AAR_BASE                      (&NRF_AAR_regs)
+extern NRF_RNG_Type NRF_RNG_regs;
+#undef NRF_RNG_BASE
+#define NRF_RNG_BASE                      (&NRF_RNG_regs)
+extern NRF_TEMP_Type NRF_TEMP_regs;
+#undef NRF_TEMP_BASE
+#define NRF_TEMP_BASE                     (&NRF_TEMP_regs)
+extern NRF_RTC_Type NRF_RTC_regs[];
+#undef NRF_RTC0_BASE
+#define NRF_RTC0_BASE                     (&NRF_RTC_regs[0])
+#undef NRF_RTC1_BASE
+#define NRF_RTC1_BASE                     (&NRF_RTC_regs[1])
+extern NRF_ECB_Type NRF_ECB_regs;
+#undef NRF_ECB_BASE
+#define NRF_ECB_BASE                      (&NRF_ECB_regs)
+extern NRF_CCM_Type NRF_CCM_regs;
+#undef NRF_CCM_BASE
+#define NRF_CCM_BASE                      (&NRF_CCM_regs)
+extern NRF_RADIO_Type NRF_RADIO_regs;
+#undef NRF_RADIO_BASE
+#define NRF_RADIO_BASE                    (&NRF_RADIO_regs)
+extern NRF_CLOCK_Type NRF_CLOCK_regs;
+#undef NRF_CLOCK_BASE
+#define NRF_CLOCK_BASE                    (&NRF_CLOCK_regs)
+extern NRF_FICR_Type NRF_FICR_regs;
+#undef NRF_FICR_BASE
+#define NRF_FICR_BASE                     (&NRF_FICR_regs)
+extern NRF_PPI_Type NRF_PPI_regs;
+#undef NRF_PPI_BASE
+#define NRF_PPI_BASE                      (&NRF_PPI_regs)
+extern NRF_TIMER_Type NRF_TIMER_regs[];
+#undef NRF_TIMER0_BASE
+#define NRF_TIMER0_BASE                   (&NRF_TIMER_regs[0])
+#undef NRF_TIMER1_BASE
+#define NRF_TIMER1_BASE                   (&NRF_TIMER_regs[1])
+#undef NRF_TIMER2_BASE
+#define NRF_TIMER2_BASE                   (&NRF_TIMER_regs[2])
+#undef NRF_TIMER3_BASE
+#define NRF_TIMER3_BASE                   (&NRF_TIMER_regs[3])
+#undef NRF_TIMER4_BASE
+#define NRF_TIMER4_BASE                   (&NRF_TIMER_regs[4])
+#undef NRF_POWER_BASE
+extern NRF_POWER_Type NRF_POWER_regs;
+#define NRF_POWER_BASE                    (&NRF_POWER_regs)
+extern NRF_GPIO_Type NRF_P0_regs;
+#undef NRF_P0_BASE
+#define NRF_P0_BASE                       (&NRF_P0_regs)
+extern NRF_GPIOTE_Type NRF_GPIOTE_regs;
+#undef NRF_GPIOTE_BASE
+#define NRF_GPIOTE_BASE                   (&NRF_GPIOTE_regs)
+extern NRF_NVMC_Type NRF_NVMC_regs;
+#undef NRF_NVMC_BASE
+#define NRF_NVMC_BASE                     (&NRF_NVMC_regs)
+
+/*
+ * Redefine the peripheral pointers
+ */
+#define NRF_FICR                    ((NRF_FICR_Type*)          NRF_FICR_BASE)
```

Comment:
How come this set needs to be defined here? (naive question)

Reply:
Some applications may access registers directly, i.e. `NRF_FICR->SOME_REGISTER` instead of using the nrfx hal functions. The redefinitions done here ensure that this statement points to a register in a mocked peripheral, and not to some invalid memory address.